cdap-app-fabric/src/main/java/io/cdap/cdap/app/DefaultAppConfigurer.java

```diff
@@ -110,7 +110,8 @@ public DefaultAppConfigurer(Id.Namespace namespace, Id.Artifact artifactId, Appl
     this.artifactId = artifactId;
     this.pluginFinder = pluginFinder;
     this.pluginInstantiator = pluginInstantiator;
-    this.appMetadata = new Metadata(Collections.emptyMap(), Collections.emptySet());
+    this.appMetadata = new HashMap<>();
+    MetadataScope.ALL.forEach(scope -> appMetadata.put(scope, new Metadata(new HashMap<>(), new HashSet<>())));
```

Comment:
instead of pre-populating with empty data, better to use computeIfAbsent() and only add it if emitMetadata() is called